pkg/blockdevice/new_block_device_from_file_windows.go

```diff
@@ -0,0 +1,95 @@
+//go:build windows
+// +build windows
+
+package blockdevice
+
+import (
+	"syscall"
+	"unsafe"
+
+	"github.com/buildbarn/bb-storage/pkg/util"
+
+	"golang.org/x/sys/windows"
+)
+
+var (
+	kernel32          = windows.NewLazySystemDLL("kernel32.dll")
+	GetDiskFreeSpaceW = kernel32.NewProc("GetDiskFreeSpaceW")
+)
+
+// NewBlockDeviceFromFile creates a BlockDevice that is backed by a
+// regular file stored in a file system.
+func NewBlockDeviceFromFile(path string, minimumSizeBytes int, zeroInitialize bool) (BlockDevice, int, int64, error) {
+	pathPtr, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return nil, 0, 0, util.StatusWrapf(err, "Failed to convert path %#v to UTF-16", path)
+	}
+
+	var createmode uint32 = windows.OPEN_ALWAYS
+	if zeroInitialize {
+		createmode = windows.CREATE_ALWAYS
+	}
+	handle, err := windows.CreateFile(
+		pathPtr,
+		windows.GENERIC_READ|windows.GENERIC_WRITE,
+		windows.FILE_SHARE_READ|windows.FILE_SHARE_WRITE,
+		nil,
+		createmode,
+		windows.FILE_ATTRIBUTE_NORMAL,
+		0)
+	if err != nil {
+		return nil, 0, 0, util.StatusWrapf(err, "Failed to open file %#v", path)
+	}
+
+	// Get the sector size from the disk where the file is located
+	var sectorSizeBytes int
+	if len(path) >= 3 && path[1] == ':' && path[2] == '\\' {
+		// Extract the root path (drive letter) from the full path
+		var rootPath string = path[:3]
+		rootPathPtr, err := syscall.UTF16PtrFromString(rootPath)
+		if err != nil {
+			windows.CloseHandle(handle)
+			return nil, 0, 0, util.StatusWrapf(err, "Failed to convert root path %#v to UTF-16", rootPath)
+		}
+
+		// Get the sector size using GetDiskFreeSpace via syscall since it's not in windows package
+		var bytesPerSector uint32
+		r, _, err := GetDiskFreeSpaceW.Call(
+			uintptr(unsafe.Pointer(rootPathPtr)),
+			0,
+			uintptr(unsafe.Pointer(&bytesPerSector)),
+			0,
+			0)
+		if r == 0 {
+			windows.CloseHandle(handle)
+			return nil, 0, 0, util.StatusWrapf(err, "Failed to get disk sector size for %#v", rootPath)
+		}
+
+		sectorSizeBytes = int(bytesPerSector)
+	} else {
+		// For UNC paths, assume a default sector size
+		sectorSizeBytes = 4096
```

Comment:
I know very little about Windows. Why doesn't the logic above work for UNC paths? Probably worth explaining. Is there some other method that could be used to determine the sector size instead? Is it even common for our users to let block devices be backed by UNC paths?

Reply:
Good point. I've had another look at that and it seems GetDiskFreeSpaceW will work for UNC paths, with some adjustments. I've written a new function to compute the "root path" that we have to pass in to GetDiskFreeSpaceW and added some unit tests for it.

One thing I wasn't sure about was how to make the utility function private whilst still adding unit tests for it so for now I've exported it. If you've got a better suggestion for this please say -- I've not written much go before so!

Reply:
Got it! Yeah, feel free to just keep it public. If it turns out we need this more frequently, then we can move it to pkg/filesystem later.